### ISSUE TITLE
RemoteConfig SQLite crash

### ISSUE THREAD

**Issue by mohammedshahid** [firebase/firebase-ios-sdk#1953]
We have integrated Firebase Remote Config , Performance in our iOS app.
With Recent **iOS 12** update the crashes started showing up crashes in Fabric with logs below.

Crash log for your reference with the particular thread that crashed : 

**Crashed: com.google.GoogleConfigService.database
EXC_BREAKPOINT 0x00000001f7bc1158**

```
#1. Crashed: com.google.GoogleConfigService.database
0  (Missing)                      0x1f7bc1158 (Missing)
1  (Missing)                      0x1f7bc1400 (Missing)
2  (Missing)                      0x33d381f84770a4 (Missing)
3  (Missing)                      0xdce01f83c8470 (Missing)
4  (Missing)                      0x183f81f83c81ac (Missing)
5  (Missing)                      0x57a781f842b710 (Missing)
6  (Missing)                      0x29b281f842f0e4 (Missing)
7  (Missing)                      0x2a9581f842e884 (Missing)
8  (Missing)                      0x3f1781f84093e0 (Missing)
9  (Missing)                      0x46c601f840250c (Missing)
10 (Missing)                      0x5918810504e978 (Missing)
11 (XXXX)                      0x10504e6b8 __77-[RCNConfigDBManager insertMainTableWithValues:fromSource:completionHandler:]_block_invoke + 4312688312
12 (Missing)                      0x1f79c6d74 (Missing)
13 (Missing)                      0x4b1d01f79c830c (Missing)
14 (Missing)                      0x122101f79cf6fc (Missing)
15 (Missing)                      0x5e6701f79d0244 (Missing)
16 (Missing)                      0x42a401f79d8524 (Missing)
17 (Missing)                      0x261e01f7bc7b38 (Missing)
18 (Missing)                      0x657b01f7bcddec (Missing)
```

Could you please update on this.

**Comment by mohammedshahid** [comment]
Pod Versions used : 
Using Firebase (5.10.0)
Using FirebaseABTesting (2.0.0)
Using FirebaseAnalytics (5.2.0)
Using FirebaseCore (5.1.5)
Using FirebaseInstanceID (3.2.2)
Using FirebasePerformance (2.1.2)
Using FirebaseRemoteConfig (3.0.2)

**Comment by mohammedshahid** [comment]
Have attached the full stack trace :

[com.cleartrip.iphoneapp_issue_7986_crash_17273e3c41224568af8202a3061b24cf_DNE_0_v2.txt](https://github.com/firebase/firebase-ios-sdk/files/2485988/com.cleartrip.iphoneapp_issue_7986_crash_17273e3c41224568af8202a3061b24cf_DNE_0_v2.txt)

**Comment by morganchen12** [comment]
Looks like this is crashing in a sqlite call we're making in RemoteConfig. Does this happen deterministically? Can you share repro steps?

**Comment by mohammedshahid** [comment]
Hey , we haven’t got this while debugging or testing . But we got a lot of crashes in production because of this issue only for iOS 12 users .
We do not have any repro steps . 

Does this have anything to do with lower Expiration Duraiton to fetch the RemoteConfig ?

**Comment by morganchen12** [comment]
Expiration duration shouldn't affect this crash.

**Comment by mohammedshahid** [comment]
Hi guys , any update on this issue .

**Comment by morganchen12** [comment]
It'll be difficult for us to make progress on this crash without repro steps.

**Comment by mohammedshahid** [comment]
Hi @paulb777 @morganchen12 - I had got this issue while debugging but its not reproducible. Please take look at the screenshots & let me know if i have to do something to fix this crash that's happening a lot for our users.

FYI, Thread 3 has crashed 

<img width="676" alt="screen shot 2018-10-24 at 5 37 58 pm" src="https://user-images.githubusercontent.com/5266699/47638315-3ec40580-db84-11e8-8545-c3bb4a863d24.png">
<img width="617" alt="screen shot 2018-10-24 at 5 38 04 pm" src="https://user-images.githubusercontent.com/5266699/47638318-3f5c9c00-db84-11e8-8ea6-f36ff32d7a04.png">
<img width="682" alt="screen shot 2018-10-24 at 5 38 08 pm" src="https://user-images.githubusercontent.com/5266699/47638321-3f5c9c00-db84-11e8-908f-e00eea165fdb.png">
<img width="686" alt="screen shot 2018-10-24 at 5 38 29 pm" src="https://user-images.githubusercontent.com/5266699/47638322-3f5c9c00-db84-11e8-9ca6-d8e50fec0c36.png">

**Comment by morganchen12** [comment]
This could potentially be a threading issue in Remote Config. @dmandar, can you investigate?

**Comment by morganchen12** [comment]
We have some pending changes in this area that should address this crash. They're part of a larger partial rewrite change, though, so it'll be some time before they're released.

In the meantime, if anyone has repro steps, we can try to verify the fixes before release.

**Comment by morganchen12** [comment]
@mohammedshahid can you share your code or any sample that reproduces this issue?

**Comment by mohammedshahid** [comment]
@morganchen12 - Hey Apologies we can't share the code and even if we do that it's hardly reproducible . We get this issue at random with no specific steps, but surely at app launch when we call - **self.firebase.setDefaults(defaultConfig as? [String : NSObject])**

**Comment by kwstasna** [comment]
We also have a crash
```
 __77-[RCNConfigDBManager insertMainTableWithValues:fromSource:completionHandler:]_block_invoke + 4301692552
--
```
as soon as the app starts. 
I think Remote Config has many flaws and its causing us some crashes also in other cases that are not relevant for this issue now.

**Comment by mohammedshahid** [comment]
@kwstasna -  I would recommend you not to use the { setDefaults(fromPlist: "") or setDefaults(fromDict: "") } method of firebase, Instead write a helper method to fetch yourself that would give you the defaults when required.

Helper in Swift :

func getDataValueForKey(_ key : String)-> Data
    {
        let configValue = self.firebase.configValue(forKey: key)
        if configValue.source == .remote {
            return configValue.dataValue
        }
        else {
            return (defaultConfig[key] as? String ?? "").data(using: .utf8)! 
//defaultConfig - is the dictionary that we maintain for the defaults.
        }
    }

**Comment by morganchen12** [comment]
@kwstasna, @mohammedshahid -- have either of you seen this crash happening recently?

**Comment by kwstasna** [comment]
@morganchen12 we see lots of crashes with the following crashlytics reports.
```
-[FPRRemoteConfigFlags update]

__42-[FPRClient startWithConfiguration:error:]_block_invoke_4

__30-[FPRRemoteConfigFlags update]_block_invoke

FIRInstanceIDAuthKeyChain.m – line 134
-[FIRInstanceIDAuthKeychain dataForService:account:]
```

Each line is a separate crash.

**Comment by morganchen12** [comment]
@kwstasna can you share the full stack trace for each crash?

**Comment by morganchen12** [comment]
Fix up internally at cr/252923920.

**Comment by morganchen12** [comment]
Fix has been merged.
